infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/KubernetesPluginsToolingApplier.java

```diff
@@ -180,8 +192,84 @@ private void addSidecar(
     InternalMachineConfig machineConfig = machineResolver.resolve();
     kubernetesEnvironment.getMachines().put(machineName, machineConfig);
 
+    sidecarRelatedCommands.forEach(c -> c.getAttributes().put("machineName", machineName));
+
+    container
+        .getCommands()
+        .stream()
+        .map(c -> asCommand(machineName, c))
+        .forEach(c -> kubernetesEnvironment.getCommands().add(c));
+
     SidecarServicesProvisioner sidecarServicesProvisioner =
         new SidecarServicesProvisioner(containerEndpoints, pod.getMetadata().getName());
     sidecarServicesProvisioner.provision(kubernetesEnvironment);
   }
+
+  private CommandImpl asCommand(String machineName, Command command) {
+    CommandImpl cmd =
+        new CommandImpl(
+            command.getName(),
+            command.getCommand().stream().collect(Collectors.joining(" ")),
+            "custom");
+    cmd.getAttributes().put("workDir", command.getWorkingDir());
+    cmd.getAttributes().put("machineName", machineName);
+    return cmd;
+  }
+
+  private static class CommandsResolver {
+    private final KubernetesEnvironment k8sEnvironment;
+    private final ArrayListMultimap<String, CommandImpl> pluginRefToCommand;
+
+    public CommandsResolver(KubernetesEnvironment k8sEnvironment) {
+      this.k8sEnvironment = k8sEnvironment;
+
+      pluginRefToCommand = ArrayListMultimap.create();
+      k8sEnvironment
+          .getCommands()
+          .forEach(
+              (c) -> {
+                String pluginRef = c.getAttributes().get("plugin");
+                if (pluginRef != null) {
+                  pluginRefToCommand.put(pluginRef, c);
+                }
+              });
+    }
+
+    private Collection<CommandImpl> resolve(ChePlugin chePlugin) {
+      List<CheContainer> containers = chePlugin.getContainers();
+
+      String pluginRef = chePlugin.getId() + ":" + chePlugin.getVersion();
+      Collection<CommandImpl> pluginsCommands = pluginRefToCommand.removeAll(pluginRef);
+
+      if (pluginsCommands.isEmpty()) {
+        // specified plugin doesn't have configured commands
+        return emptyList();
+      }
+
+      if (containers.isEmpty()) {
+        k8sEnvironment
+            .getWarnings()
+            .add(
+                new WarningImpl(
+                    44012,
+                    format(
+                        "There are configured commands for plugin '%s' that doesn't have any containers",
+                        pluginRef)));
+        return emptyList();
+      }
+
+      if (containers.size() > 1) {
+        k8sEnvironment
+            .getWarnings()
+            .add(
+                new WarningImpl(
+                    44013,
```

Comment:
Can you elaborate on why did you choose these magic numbers for error codes>

Reply:
As far as I know, we don't have a clear way of using and introducing warnings codes, do we?. So, this process is magic for me. I choose a number which I like ;)

Reply:
Please take a look the following commit https://github.com/eclipse/che/pull/12016/commits/4f45ebbd7618b17f9e9687961e852cd47069ad3e which introduces `Warnings` class for holdings warnings codes and warnings.